### PR TITLE
Add support for currency on Expense items

### PR DESCRIPTION
**Pull request by Betree**
API part for https://github.com/opencollective/opencollective/issues/7010
Related to https://github.com/opencollective/opencollective/issues/6906
Resolve https://github.com/opencollective/opencollective/issues/6588
Resolve https://github.com/opencollective/opencollective/issues/7149

### CODE REVIEW

File: server/lib/currency.ts [hunk 1]
```diff
@@ -150,13 +163,13 @@ export async function fetchFxRates(
...
-  if (!isLiveEnv) {
+  if (useMockRate) {
     const ratesValues = toCurrencies.map(() => 1.1);
     return zipObject(toCurrencies, ratesValues);
-  } else {
-    return getRatesFromDb(fromCurrency, toCurrencies, date);
   }
+
+  // As a fallback, try to fetch the rates from the DB
+  return getRatesFromDb(fromCurrency, toCurrencies, date);
```
Comment (kewitz) [inline server/lib/currency.ts:172]: 🍨 A possible optimization would be checking the DB first for a fresh enough date and storing the recently fetched rate from Fixer every time we don't find it.

**Review by kewitz — COMMENTED** [review]
Partial Review.

File: migrations/20231124101312-add-currency-to-expense-items.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+module.exports = {
+  async up(queryInterface, Sequelize) {
+    // Create new columns
+    await queryInterface.addColumn('ExpenseItems', 'currency', {
+      type: Sequelize.STRING(3),
+      allowNull: false,
+      defaultValue: '___', // We're doing that rather than setting to null then enforcing NOT NULL because we could face problems if expenses get updated between the migration & the deployment. Also, it's probably faster.
+    });
+
+    await queryInterface.addColumn('ExpenseItems', 'fxRate', {
```
[the remark this answers is not in the document]
Comment (Betree) [reply migrations/20231124101312-add-currency-to-expense-items.js:13]: `itemToExpenseFxRate`

File: server/graphql/common/expenses.ts [hunk 3]
```diff
@@ -1220,17 +1207,98 @@ const checkCanUseAccountingCategory = (
   }
 };
 
+export const prepareExpenseItemInputs = async (
+  expenseCurrency: string,
+  itemsInput: Array<ExpenseItem | Record<string, unknown>>,
```
[the remark this answers is not in the document]
Comment (Betree) [reply server/graphql/common/expenses.ts:1212]: We depend on https://github.com/opencollective/opencollective-api/blob/db00b65dde798994823101738e0a374a71d93f05/server/graphql/common/expenses.ts#L1143

File: migrations/20231124101312-add-currency-to-expense-items.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      type: Sequelize.FLOAT,
+      allowNull: true,
```
Comment (Betree) [inline migrations/20231124101312-add-currency-to-expense-items.js:15]:
```suggestion
      allowNull: false,
      defaultValue: 1
```

File: server/graphql/common/expenses.ts [hunk 5]
```diff
@@ -1872,12 +1945,22 @@ export async function editExpense(req: express.Request, expenseData: ExpenseData
           return item.destroy({ transaction });
         }),
         // Create
+        // TODO items need to be prepared here
```
Comment (Betree) [inline server/graphql/common/expenses.ts:1948]: Remove this todo, it's done

File: server/graphql/common/expenses.ts [hunk 6]
```diff
@@ -1872,12 +1945,22 @@ export async function editExpense(req: express.Request, expenseData: ExpenseData
...
-          return models.ExpenseItem.createFromData(itemData, remoteUser, expense, transaction);
+          return models.ExpenseItem.createFromData(
+            { ...itemData, currency: itemData.currency || expenseData.currency || expense.currency },
+            remoteUser,
+            expense,
+            transaction,
+          );
         }),
         // Update
+        // TODO items need to be prepared here
```
Comment (Betree) [inline server/graphql/common/expenses.ts:1958]: Remove this todo, it's done

File: server/lib/currency.ts [hunk 7]
```diff
@@ -33,9 +35,19 @@ export function formatCurrency(currency: string, value: number): string {
...
 }
 
-if (!get(config, 'fixer.accessKey') && !['staging', 'production'].includes(config.env)) {
-  logger.info('Fixer API is not configured, lib/currency will always return 1.1');
-}
+const showFixerWarning = () => {
+  if (!get(config, 'fixer.accessKey')) {
+    if (['staging', 'production'].includes(config.env) || parseToBoolean(get(config, 'fixer.disableMock'))) {
+      const logFn = config.env === 'production' ? logger.warn : logger.info; // In prod, having no Fixer API would be problematic as our FX rates would quickly become outdated
+      logFn('Fixer API is not configured, lib/currency will fallback to DB values');
```
[the remark this answers is not in the document]
Comment (Betree) [reply server/lib/currency.ts:42]: > Hmm, this condition is quite dangerous. 
Before this change, there would have been no logs at boot time in production if Fixer was not configured. This condition actually adds an error log in production that didn't exist before.

File: migrations/20231124101312-add-currency-to-expense-items.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Betree) [reply migrations/20231124101312-add-currency-to-expense-items.js:13]: For consistency with Transaction's `hostCurrencyFxRate`, I'll actually go with `expenseCurrencyFxRate`

**Pull request merged**